### ISSUE TITLE
Crash when viewing malformed Applesoft BASIC file

### ISSUE THREAD

**Issue by ryandesign** [AppleCommander/AppleCommander#119]
I remembered that AppleCommander supports converting between tokenized Applesoft BASIC and a text representation but not exactly how to accomplish it.

Fumbling my way through, with AppleCommander 1.8.0 on macOS Monterey, I did this:

1. In the main window, click Create… in the toolbar and use the Create Disk Wizard to make a new DOS disk.
2. In the disk window, click Import… in the toolbar and in the Import Wizard choose a text file that contains an Applesoft BASIC program.
3. Its type was autodetected as binary (B) but since I wanted AppleCommander to convert it to tokenized Applesoft BASIC I selected it, chose Edit…, and changed the Filetype menu from B to A and clicked OK.
4. After clicking Finish, the file was added to the disk image.

Because there were no warnings or errors I believed this had done what I wanted, but instead what I think happened is that the file's data was copied byte for byte to the disk image and the file's type was simply set to A without its contents having actually been converted to that format. This seems confirmed by the fact that the file on the disk image is the same size as the original; if it had been tokenized it should be smaller.

One reason why I thought these steps were correct is that there is a "RAW binary?" checkbox in the File Import Settings window. I assumed that checking this box would copy the data verbatim. Since I did not check this box, I assumed a conversion would take place.

Then, selecting the file in the listing, I clicked View in the toolbar, which caused AppleCommander to quit; this is what was in the terminal:

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 61 out of bounds for length 61
	at com.webcodepro.applecommander.util.ApplesoftTokenizer.getNextToken(ApplesoftTokenizer.java:129)
	at com.webcodepro.applecommander.ui.swt.filteradapter.ApplesoftFilterAdapter.createStyledText(ApplesoftFilterAdapter.java:74)
	at com.webcodepro.applecommander.ui.swt.filteradapter.ApplesoftFilterAdapter.display(ApplesoftFilterAdapter.java:47)
	at com.webcodepro.applecommander.ui.swt.FileViewerWindow.open(FileViewerWindow.java:168)
	at com.webcodepro.applecommander.ui.swt.DiskExplorerTab.viewFile(DiskExplorerTab.java:1631)
	at com.webcodepro.applecommander.ui.swt.DiskExplorerTab$52.widgetSelected(DiskExplorerTab.java:1379)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4565)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1545)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1530)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4351)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3927)
	at com.webcodepro.applecommander.ui.swt.SwtAppleCommander.launch(SwtAppleCommander.java:118)
	at com.webcodepro.applecommander.ui.swt.SwtAppleCommander.launch(SwtAppleCommander.java:86)
	at com.webcodepro.applecommander.ui.swt.SwtAppleCommander.main(SwtAppleCommander.java:71)
	... 8 more
```

So I think the bug is just that the Applesoft BASIC viewer expects an Applesoft BASIC file to be completely valid. Some error checking should be added so that if for example invalid tokens are encountered an error message is displayed and the program can continue normally rather than exiting abnormally.

**Comment by a2geek** [comment]
There is some capability to import Applesoft BASIC files from a text file -- but it's in the command-line tools and not in the user interface itself. If you're curious, `ac` has the `-bas` flag, and a sample would be like `cat ~/a.bas | ac -bas cc65.dsk test1`. This will take the text and tokenize it as expected. `acx import` has some flags to toggle this as well. And, finally, there is the "basic tokenizer" (`bt`) tool that does a little bit more. Nothing super special, but they may be useful.

For doc references:
* `ac` - https://applecommander.github.io/ac/#put-standard-input-basic-source-code-onto-disk-image-as-a-basic-file
* `acx` - the first sample shows an import as well - https://github.com/AppleCommander/AppleCommander/tree/main/app/cli-acx
* `bt` is part of bastools - https://github.com/AppleCommander/bastools

What I think I'll do is look at validating that the input file is actually a BASIC program and not something else.

The UI is 20+ years old (seriously) and definitely more on the clunky side at times. :man_shrugging:

**Comment by ryandesign** [comment]
Right, I'm aware of the command line support for tokenizing Applesoft for import. It's been a couple years since I've played with AppleCommander and I didn't remember the capabilities exactly, plus I had just updated from 1.6.0 to 1.8.0 and didn't know what had changed between, so I was just playing around with the UI when I noticed this issue.

Validating Applesoft before import is a good idea. But not crashing when viewing an invalid file is also a good idea. I'm imagining that invalid files could result any number of ways, such as attempting to read a corrupted disk.

I tried rebuilding 1.8.0 with 709ce6b47f17ab951b327b1bc07f66bf5b8cf8ab added and now when I go to the Import Wizard, click Choose…, and select some file, it says "File does not appear to be an Applesoft program". That's true: the file I selected was not Applesoft. But what if I wanted to import it as a binary file?

Not all files do this. I was able to select a text file for import. It was auto-detected as filetype T. If I select the text file in the list, click Edit…, change its filetype to A, click OK, and click Finish, it is imported and its filetype is set to Applesoft. If I select that file in the disk window and click View, AppleCommander still crashes.

So what is the purpose of the "RAW binary?" checkbox anyway?
